src/Servant/Multipart.hs

```diff
@@ -308,36 +309,46 @@ fuzzyMultipartCTCheck ct
 --   the temporary file backend. See haddocks for
 --   'ParseRequestBodyOptions' and 'TmpBackendOptions' respectively
 --   for more information on what you can tweak.
-data MultipartOptions options = MultipartOptions
+data MultipartOptions tag = MultipartOptions
   { generalOptions :: ParseRequestBodyOptions
-  , options        :: options
+  , options        :: MultipartBackendOptions tag
   }
 
-class MultipartBackend options where
-    type MultipartResult options :: *
+class MultipartBackend tag where
+    type MultipartResult tag :: *
+    type MultipartBackendOptions tag :: *
 
-    backend :: options
+    backend :: Proxy tag
+            -> MultipartBackendOptions tag
             -> InternalState
             -> ignored1
             -> ignored2
             -> IO SBS.ByteString
-            -> IO (MultipartResult options)
+            -> IO (MultipartResult tag)
 
-    defaultBackendOptions :: options
+    defaultBackendOptions :: Proxy tag -> MultipartBackendOptions tag
 
-instance MultipartBackend TmpBackendOptions where
-    type MultipartResult TmpBackendOptions = FilePath
+-- | Tag for temporary files
+data Tmp
 
-    defaultBackendOptions = defaultTmpBackendOptions
-    backend opts = tmpBackend
+-- | Tag for items in memory
+data Mem
+
+instance MultipartBackend Tmp where
+    type MultipartResult Tmp = FilePath
+    type MultipartBackendOptions Tmp = TmpBackendOptions
+
+    defaultBackendOptions _ = defaultTmpBackendOptions
+    backend _ opts = tmpBackend
       where
         tmpBackend = tempFileBackEndOpts (getTmpDir opts) (filenamePat opts)
 
-instance MultipartBackend LbsBackendOptions where
-    type MultipartResult LbsBackendOptions = LBS.ByteString
+instance MultipartBackend Mem where
+    type MultipartResult Mem = LBS.ByteString
+    type MultipartBackendOptions Mem = LbsBackendOptions
```

Comment:
Well, `LbsBackendOptions` is just a placeholder for now right? There's going to be something fleshed out by the time this PR is ready isn't there?

If there is going to be something, I'd suggest leaving it as is. Otherwise (i.e if we want to go with an option-less in-memory backend for the firest release of this backend), I'd suggest just using `()` indeed.